ddl/ddl_tiflash_api.go

```diff
@@ -0,0 +1,512 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2013 The ql Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSES/QL-LICENSE file.
+
+package ddl
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/ddl/placement"
+	ddlutil "github.com/pingcap/tidb/ddl/util"
+	"github.com/pingcap/tidb/domain/infosync"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/store/helper"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/admin"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/gcutil"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// TiFlashReplicaStatus records status for each TiFlash replica.
+type TiFlashReplicaStatus struct {
+	ID             int64
+	Count          uint64
+	LocationLabels []string
+	Available      bool
+	HighPriority   bool
+}
+
+// TiFlashManagementContext is the context for TiFlash Replica Management
+type TiFlashManagementContext struct {
+	TiFlashStores             map[int64]helper.StoreStat
+	HandlePdCounter           int
+	UpdateTiFlashStoreCounter int
+	UpdateMap                 map[int64]bool
+}
+
+// NewTiFlashManagementContext creates an instance for TiFlashManagementContext.
+func NewTiFlashManagementContext() *TiFlashManagementContext {
+	return &TiFlashManagementContext{
+		HandlePdCounter:           0,
+		UpdateTiFlashStoreCounter: 0,
+		TiFlashStores:             make(map[int64]helper.StoreStat),
+		UpdateMap:                 make(map[int64]bool),
+	}
+}
+
+var (
+	// PollTiFlashInterval is the interval between every PollTiFlashReplicaStatus call.
+	PollTiFlashInterval = 1 * time.Second
+	// PullTiFlashPdTick indicates the number of intervals before we fully sync all TiFlash pd rules and tables.
+	PullTiFlashPdTick = 60 * 5
+	// UpdateTiFlashStoreTick indicates the number of intervals before we fully update TiFlash stores.
+	UpdateTiFlashStoreTick = 10
+	// ReschePullTiFlash is set true, so we do a fully sync, regardless of PullTiFlashPdTick.
+	// Set to be true, when last TiFlash pd rule fails.
+	ReschePullTiFlash = uint32(0)
+)
+
+func makeBaseRule() placement.Rule {
+	return placement.Rule{
+		GroupID:  "tiflash",
+		ID:       "",
+		Index:    0,
+		Override: true,
```

Comment:
Are you suggesting that we can set index to a very high value, and override to true, so that this rule can't be override by other group, and will not override other group's config?